src/python/pants/jvm/target_types.py

```diff
@@ -4,17 +4,48 @@
 from __future__ import annotations
 
 from abc import ABCMeta
+from typing import Optional
 
+from pants.engine.addresses import Address
 from pants.engine.target import (
     COMMON_TARGET_FIELDS,
     FieldSet,
+    InvalidFieldException,
     SingleSourceField,
     StringField,
     StringSequenceField,
     Target,
 )
 from pants.util.docutil import git_url
 
+# -----------------------------------------------------------------------------------------------
+# Generic resolve support fields
+# -----------------------------------------------------------------------------------------------
```

Comment:
This section is moved, unmodified, from below. I'll need it in a followup PR defined here so that I can subclass it lower in the file.